src/category_theory/limits/shapes/cartesian_closed.lean

```diff
@@ -0,0 +1,387 @@
+class exponentiable {C : Type u} [category.{v} C] [has_finite_products.{v} C] (X : C) :=
+(is_adj : is_left_adjoint (prod_functor.obj X))
+
+/--
+If `X` and `Y` are exponentiable then `X ⨯ Y` is.
+This isn't an instance because it's not usually how we want to construct exponentials
+-/
+def binary_product_exponentiable {C : Type u} [category.{v} C] [has_finite_products.{v} C] {X Y : C}
+  (hX : exponentiable X) (hY : exponentiable Y) : exponentiable (X ⨯ Y) :=
+{ is_adj :=
+  by haveI := hX.is_adj; haveI := hY.is_adj;
+     exact adjunction.left_adjoint_of_nat_iso (prod_functor_left_comp _ _).symm }
+
+/-- A category `C` is cartesian closed if every object is exponentiable, and it has finite prodcuts. -/
+class is_cartesian_closed (C : Type u) [category.{v} C] [has_finite_products.{v} C] :=
+(cart_closed : Π (X : C), exponentiable X)
+
+attribute [instance, priority 100] is_cartesian_closed.cart_closed
+
+variables {C : Type u} [category.{v} C] (A B : C) {X X' Y Y' Z : C}
+
+section exp
+variables [has_finite_products.{v} C] [exponentiable A]
+
+/-- This is (-)^A. -/
+def exp.functor : C ⥤ C :=
+(@exponentiable.is_adj _ _ _ A _).right
+
+/-- The adjunction between A ⨯ - and (-)^A. -/
+def exp.adjunction : prod_functor.obj A ⊣ exp.functor A :=
+exponentiable.is_adj.adj
+
+/-- The evaluation natural transformation. -/
+def ev.nat_trans : exp.functor A ⋙ prod_functor.obj A ⟶ 𝟭 C :=
+exponentiable.is_adj.adj.counit
+
+/-- The coevaluation natural transformation. -/
+def coev.nat_trans : 𝟭 C ⟶ prod_functor.obj A ⋙ exp.functor A :=
+exponentiable.is_adj.adj.unit
+
+/-- `B ^ A` or `A ⟹ B` -/
+def exp : C := (exp.functor A).obj B
+
+infixl `⟹`:20 := exp
+infixr `^^`:30 := pow
+
+/-- Postcompose an internal hom with an external hom. -/
+def post (f : X ⟶ Y) : A⟹X ⟶ A⟹Y :=
+(exp.functor A).map f
+
+/-- Postcomposition of a composition decomposes. -/
+lemma post.map_comp {f : X ⟶ Y} {g : Y ⟶ Z} : post A (f ≫ g) = post A f ≫ post A g :=
+(exp.functor A).map_comp _ _
+
+/-- The evaluation morphism. -/
+def ev : A ⨯ (A⟹B) ⟶ B :=
+(ev.nat_trans A).app B
+
+/-- The coevaluation morphism. -/
+def coev : B ⟶ A⟹(A⨯B) :=
+(coev.nat_trans A).app B
+
+@[simp, reassoc] lemma ev_coev : limits.prod.map (𝟙 A) (coev A B) ≫ ev A (A ⨯ B) = 𝟙 (A ⨯ B) :=
+adjunction.left_triangle_components (exp.adjunction A)
+
+@[simp, reassoc] lemma coev_ev : coev A (A⟹B) ≫ post A (ev A B) = 𝟙 (A⟹B) :=
+adjunction.right_triangle_components (exp.adjunction A)
+
+/-- Coevaluation is natural. -/
+@[simp, reassoc, priority 10]
+lemma coev_nat (f : X ⟶ Y) : f ≫ coev A Y = coev A X ≫ post A (limits.prod.map (𝟙 A) f) :=
+(coev.nat_trans A).naturality f
+
+/-- Evaluation is natural. -/
+@[simp, reassoc, priority 10]
+lemma ev_nat (f : X ⟶ Y) : limits.prod.map (𝟙 A) (post _ f) ≫ ev A Y = ev A X ≫ f :=
+(ev.nat_trans A).naturality f
+
+end exp
+
+variables {A}
+
+-- Wrap these in a namespace so we don't clash with the core versions.
+namespace cart_closed
```

Comment:
Should be called `is_cartesian_closed`. It's not a good idea to have random namespace names. :-)